File: 11.sorting/bubble_sort.py
class BubbleSort:
    def sort(self, input: list[str]) -> list[str]:
        for i in range(len(input)):
            for j in range(len(input) - 1, i, -1):
                # right < left
                if input[j] < input[j-1]:
                    input[j-1], input[j] = input[j], input[j-1]
        return input


b = BubbleSort()

input = ["b", "c", "d", "a", "e", "f", "z", "y", "w", "h", "g"]
print(f"Bubble Sort Result: {b.sort(input)}")

input = [5, 8, 3, 9, 4, 1, 7]
print(f"Bubble Sort Result: {b.sort(input)}")

File: 11.sorting/bubble_sort.go
package main

/**
func BubbleSort(input []string) []string {
	for i from 0 to n-1 {
		for j in n-1 down to i+1 {
			if input[j-1] > input[j] {
				Swap: input[j-1], input[j]
			}
		}
	}
	return input
}
*/

func BubbleSort(input []string) []string {
	for i := range input {
		for j := len(input) - 1; j > i; j -= 1 {
			if input[j] < input[j-1] {
				input[j], input[j-1] = input[j-1], input[j]
			}
		}
	}
	return input
}

// func main() {
// 	input := []string{"b", "c", "d", "a", "a", "e", "f", "z", "a", "y", "w", "h", "g", "y"}

// 	sorted := BubbleSort(input)

// 	fmt.Printf("Bubble Sort Result: %v\n", sorted)
// }
